### PR TITLE
Equalize promoted

### PR DESCRIPTION
**Pull request by PeterSolMS**
I have observed that in server GC scenarios, the amount of promoted memory is often very uneven between heaps, which leads to suboptimal work distribution between server GC threads.

The PR introduces a new method equalize_promoted_bytes that attemps to even out the promoted memory between heaps by moving regions from heaps with lots of promotion of heaps with less promotion.

The algorithm used removes regions from heaps that have more than average promotion. These surplus regions are arranged into size classes by the amount of promoted memory in them. The heaps are also arranged into size classes by how much their promoted memory is short of the average promoted memory per heap.

Then we repeatedly move the surplus region with the most promoted memory to the heap with the biggest deficit. If the heap still has a deficit, it is reconsidered under its new deficit size class. Otherwise, it now has average or more promoted memory and is removed from further consideration.

Because regions may now move between heaps, it is no longer true that the finalization queue entry for an object is always on the same heap as the object itself. This necessitated moving the call to finalize_queue->UpdatePromotedGenerations to a later point in time when all threads have finished updating the final generation for a region.

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -22143,6 +22144,433 @@ void gc_heap::sync_promoted_bytes()
...
+        if (avg_surv_per_heap != 0)
+        {
+            dprintf (REGIONS_LOG, ("before equalize: gen: %d avg surv: %Id max_surv: %Id imbalance: %d", gen_idx, avg_surv_per_heap, max_surv_per_heap, max_surv_per_heap*100/avg_surv_per_heap));
+        }
+#if 0 // see code for more sophisticated approach below
+        // step 2:
+        int deficit_heaps[MAX_SUPPORTED_CPUS];
+        int num_deficit_heaps = 0;
+        int surplus_heaps[MAX_SUPPORTED_CPUS];
+        int num_surplus_heaps = 0;
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:22300]: you might want to make `deficit_heaps` and `surplus_heaps` arrays of `short`s. we are taking quite a bit of stack space with these ararys..

File: src/coreclr/gc/gc.cpp [hunk 2]
```diff
@@ -22143,6 +22144,433 @@ void gc_heap::sync_promoted_bytes()
...
+                //  dipping below the average, go into surplus heaps
+                heap_segment* first_region = g_heaps[i]->peek_first_rw_region (gen_idx);
+                assert (surv_per_heap[i] >= heap_segment_survived (first_region));
+                if (surv_per_heap[i] - heap_segment_survived (first_region) >= avg_surv_per_heap)
+                {
+                    assert (num_surplus_heaps < MAX_SUPPORTED_CPUS-1);
+                    surplus_heaps[num_surplus_heaps++] = i;
+                }
+            }
+            //  all other heaps are not looked at further
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:22324]: not sure what this comment means

File: src/coreclr/gc/gc.cpp [hunk 3]
```diff
@@ -22143,6 +22144,433 @@ void gc_heap::sync_promoted_bytes()
...
+        }
+
+        // step 3:
+        //  as long as we have surplus heaps and deficit heaps,
+        //  move regions from surplus heaps to deficit heaps
+        while (num_surplus_heaps > 0 && num_deficit_heaps > 0)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:22330]: nit - formatting

        `while ((num_surplus_heaps > 0) && (num_deficit_heaps > 0))`

File: src/coreclr/gc/gc.cpp [hunk 4]
```diff
@@ -22143,6 +22144,433 @@ void gc_heap::sync_promoted_bytes()
...
+    uint8_t* region_end = heap_segment_reserved (region);
+
+    int num_basic_regions = (int)((region_end - region_start) >> min_segment_size_shr);
+    for (int i = 0; i < num_basic_regions; i++)
+    {
+        uint8_t* basic_region_start = region_start + ((size_t)i << min_segment_size_shr);
+        heap_segment* basic_region = get_region_info (basic_region_start);
+        heap_segment_heap (basic_region) = nullptr;
+    }
+
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:22214]: would make sense to make a little util method of this as it's duplicated in `thread_rw_region_front`...

**Comment by Maoni0** [comment]
I do think the more sophisticated way is worthwhile so you can ignore my comments in the `#if 0` block.

File: src/coreclr/gc/gc.cpp (hunk 4 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:22214]: Good point, I made a method set_heap_for_contained_basic_regions that sets the heap_segment_heap for the contained basic regions.

**Review by Maoni0 — APPROVED** [review]
LGTM! the only minor thing if you don't actually need this dprintf it'd be good to get rid of it since it isn't related to this feature -

`dprintf (3, ("uoh alloc: could not release lock on %Ix", obj));`

**Pull request merged**